pkg/tests/end_to_end_tests/drop_test.go

```diff
@@ -207,6 +209,110 @@ func TestSQLDropChunk(t *testing.T) {
 }
 
+// TestSQLDropChunkWithLocked tests the case where some metrics are locked in the
+// first loop of execute_data_retention_policy
+func TestSQLDropChunkWithLocked(t *testing.T) {
+	if testing.Short() {
+		t.Skip("skipping integration test")
+	}
+	if !*useTimescaleDB {
+		t.Skip("This test only runs on installs with TimescaleDB")
+	}
+	withDB(t, *testDatabase, func(db *pgxpool.Pool, t testing.TB) {
+		dbJob := testhelpers.PgxPoolWithRole(t, *testDatabase, "prom_maintenance")
+		defer dbJob.Close()
+		//a chunk way back in 2009
+		oldChunk1 := time.Date(2009, time.November, 11, 0, 0, 0, 0, time.UTC)
+
+		ts := []prompb.TimeSeries{
+			{
+				Labels: []prompb.Label{
+					{Name: pgmodel.MetricNameLabelName, Value: "test"},
+					{Name: "name1", Value: "value1"},
+				},
+				Samples: []prompb.Sample{
+					{Timestamp: int64(model.TimeFromUnixNano(oldChunk1.UnixNano()) - 1), Value: 0.1},
+					{Timestamp: int64(model.TimeFromUnixNano(time.Now().UnixNano()) - 1), Value: 0.1},
+				},
+			},
+			{
+				Labels: []prompb.Label{
+					{Name: pgmodel.MetricNameLabelName, Value: "test2"},
+					{Name: "name1", Value: "value1"},
+				},
+				Samples: []prompb.Sample{
+					{Timestamp: int64(model.TimeFromUnixNano(time.Now().UnixNano()) - 1), Value: 0.1},
+				},
+			},
+		}
+		ingestor, err := ingstr.NewPgxIngestorForTests(pgxconn.NewPgxConn(db), nil)
+		if err != nil {
+			t.Fatal(err)
+		}
+		defer ingestor.Close()
+		_, _, err = ingestor.Ingest(newWriteRequestWithTs(copyMetrics(ts)))
+		if err != nil {
+			t.Error(err)
+		}
+
+		var tableName string
+		var metricId int
+		err = db.QueryRow(context.Background(), "SELECT id, table_name FROM _prom_catalog.get_metric_table_name_if_exists('test')").Scan(&metricId, &tableName)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		cnt := 0
+		err = db.QueryRow(context.Background(), fmt.Sprintf(`SELECT count(*) FROM show_chunks('prom_data."%s"')`, tableName)).Scan(&cnt)
+		if err != nil {
+			t.Fatal(err)
+		}
+		if cnt != 2 {
+			t.Errorf("Expected there to be a chunk")
+		}
+
+		//take the lock before executing execute_maintenance
+		tx, err := db.Begin(context.Background())
+		require.NoError(t, err)
+		_, err = tx.Exec(context.Background(), "SELECT _prom_catalog.lock_metric_for_maintenance($1)", metricId)
+		require.NoError(t, err)
+
+		wg := sync.WaitGroup{}
+		wg.Add(2)
+
+		go func() {
+			defer wg.Done()
+			//this will block until the lock is released
+			_, err = dbJob.Exec(context.Background(), "CALL prom_api.execute_maintenance(log_verbose=>true)")
+			if err != nil {
+				t.Fatal(err)
+			}
+		}()
+
+		//give it a sec to go through the first loop
+		time.Sleep(time.Second)
+
+		go func() {
+			defer wg.Done()
+			//release the lock to allow execute maintenance to complete
+			_, err := tx.Exec(context.Background(), "SELECT _prom_catalog.unlock_metric_for_maintenance($1)", metricId)
+			require.NoError(t, err)
+			err = tx.Commit(context.Background())
+			require.NoError(t, err)
+		}()
+
+		wg.Wait()
```

Comment:
I don't really like using `time.Sleep`  to synchronize goroutines for a couple of reasons but mostly because it reminds me of Javascript :grin:. 

I suggest removing the wait group and sleep call. Just call the release lock in a separate goroutine and execute maintenance in the main goroutine. That should block until release goroutine manages to catch up.

This way we are guaranteed to run exec maintenance before releasing the lock and we don't need to wait around.